Implement user symbol translation for BPFd

jcanseco opened the pull request.

jcanseco committed:
Implement commands for user symbol translation

This allows users of BPFd to determine the user symbol name that
corresponds to a given memory address (and vice versa) for a remote
device.

Signed-off-by: Jazel Canseco <jcanseco@google.com>

jcanseco commented:
This depends on PR #25.

joelagnel commented:
I don't see the point of the extra `SymbolResolver` class really. The centralization argument is too weak to warrant creation of a new C++ source file with additional caching logic, and the cache creation can be done only once per-pid from bpfd.c itself, instead of creating it every time on every lookup command.

joelagnel commented on src/bpfd.c:
```diff
@@ -700,6 +700,50 @@ int main(int argc, char **argv)
+			PARSE_UINT64(addr);
+			PARSE_INT(demangle);
+
+			usym_cache = bcc_symcache_new(pid, NULL);
```
Instead of introducing an issue (creation of cache every time), and fixing it in a later commit, you should design it correctly to begin with.

jcanseco commented:
Hey, so I wasn't 100% sure if the SymbolResolver class was the way to go, which is why I added it as a separate commit for now, so that I can easily reverse it if you did not think it was a good idea.

The issue of having an extra commit aside though, how can we do the caching of the ProcSyms instances just from bpfd.c? I basically introduced SymbolResolver so we can take advantage of C++'s std::unordered_map as a nice way to cache ProcSyms for each pid (so that we don't recreate it for every sym lookup for each pid), but I'm kind of confused how we can efficiently replicate this functionality from just C in bpfd.c alone since C doesn't really come with any nice data structures aside from, well, arrays. Am I just missing something here?

joelagnel commented:
Just create an array of PID->pointers

jcanseco commented:
Okay, so what kind of implementation are looking at here?

I'm assuming we're thinking of something like this:

We're gonna use an array as our main data structure for holding references to the ProcSyms instances.

The array itself can be an array of usym structs, where a usym struct can be defined as:

```
struct usym {
    int pid;
    void *usym_cache;
}
```
This way we can retrieve the ProcSyms instance that belongs to a given PID. We can also find out for which PIDs we've previously created a ProcSyms instance for.

The problem I see with this though is how do we handle the case where the array becomes filled to capacity, and we have to create another ProcSyms instance? I understand we can just create a new array and copy all the elements over, but this'll be just somewhat of a re-implementation of std::vector.

Also, PID lookup in this case would be O(n) since we would have to traverse the whole array in the worst-case scenario to find the particular PID we're looking for, as opposed to std::unordered_map's O(1) lookup. Wouldn't this hamper performance since some BCC tools perform _a lot_ of symbol lookups?

I'm still somewhat of a beginner at C and my background is mostly OOP, so please forgive me if this is actually just an obvious problem.

joelagnel commented:
You can index the array with the PID which makes look up O(1). 32768 is typically maximum pids in Linux, but if a rare collision occurs (say max pids run out or a pid is reused), then a collision can be detected easily by storing the full name of the exe along with the PID in the array. To handle the case of number of PIDS > 32k, which isn't the default - you can easily just store the full PID and the full EXE name and detect collisions. Incase of collisions, just return "unknown". I'd say worry about these cases later though and for now just do a plain PID->ProcSyms(or whatever ptr) array mapping.

joelagnel commented:
You're not handling collisions or the special cases in your earlier PR
anyway by the way...

On Fri, Mar 9, 2018 at 11:35 AM, Jazel Canseco <notifications@github.com>
wrote:

> Hey, so I wasn't 100% sure if the SymbolResolver class was the way to go,
> which is why I added it as a separate commit for now, so that I can easily
> reverse it if you did not think it was a good idea.
>
> The issue of having an extra commit aside though, how can we do the
> caching of the ProcSyms instances just from bpfd.c? I basically introduced
> SymbolResolver so we can take advantage of the C++'s std::unordered_map as
> a nice way to cache ProcSyms for each pid (so that we don't recreate it for
> every sym lookup for each pid), but I'm kind of confused how we can
> efficiently replicate this functionality from just C in bpfd.c alone. Am I
> just missing something here?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joelagnel/bpfd/pull/26#issuecomment-371922806>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACSVA7dlGkKEm5UjXX__-tbDdxNpIA1ks5tctmFgaJpZM4Sjk5u>
> .
>

jcanseco commented:
Ah okay that explains it then. I didn't know there was a max cap for PIDs in Linux. I'll deal with this problem in a bit then.

joelagnel commented:
Default is 32k, so that's why I mentioned to handle collisions. If a collision occurs, its just fair to make the request fail for now.

So the array should look like:
Array[PID % 32k] = struct { FULL-PID, POINTER-TO-ProcSyms }

Compare FULL-PID on a look up and make sure, its the same as expected. If its different, then there's a collision and make the look up fail. Add some useful messages that a collision occurred.

jcanseco commented:
For the collisions scenario, isn't it enough to just return a failure if the given PID goes over the 32k default cap? (i.e. no need to check if a ProcSyms already exists at the index PID % 32k, but just do a simple PID > 32k check)

For the other type of collision (where the PID is reused for a different process), wouldn't it be better to just overwrite the existing ProcSyms we cached before for that PID? After all, the PID->ProcSyms mapping held at that index is now outdated. Also, this seems better to me since then we can still return a successful lookup instead of just outright failing the request. Currently thinking about how to implement this, but didn't get to it yet, however. It's also worth nothing that BCC does not handle this collision case either for some reason.

jcanseco commented:
Also please see the most recent change I made. I'll squash it later into the first commit; I just want to bring your attention to the changes I made related to caching ProcSyms to see if it's suitable. 

Naturally, I already made sure to test it.

jcanseco commented:
Added another commit that checks for the PID going over the default max of 32k. I also plan to squash this with the first commit ("Implement user symbol translation for BPFd")

joelagnel commented:
No. If PIDs are > 32k and nothing less than 32k was cached, then it should be cached since there's nothing to collide with. In other words its normal and not an error.

jcanseco committed:
Add caching mechanism to cache ProcSyms instances

This is to avoid the needless recreation of ProcSyms every time a
userspace symbol is looked up.

The key-value caching mechanism, which uses pid's as keys, also handles
the following pid collision cases: (1) a pid is reused by a new,
different process; and (2) the pid of a process goes over the default max
pid for Linux machines.

Signed-off-by: Jazel Canseco <jcanseco@google.com>

jcanseco commented:
Okay, how about this one (see most recent commit).

It caches the ProcSyms instance, and handles both collision cases (pid reuse and pid wrap-around due to 32k cap).

Let me know what you think. If you think it's good, I'll squash the commits and propagate the change onto the [BCC PR](https://github.com/joelagnel/bcc/pull/8).

joelagnel commented on src/bpfd.c:
```diff
@@ -331,12 +339,54 @@ int bpf_clear_map(int map_fd, int klen)
 	char *cmd, *lineptr, *argstr, *tok, *kvers_str = NULL;
 	int len, c, kvers = -1;
 	void *ksym_cache = NULL;
+	struct usym_cache *usym_caches[DEFAULT_MAX_PID] = { NULL };
```
Dynamically allocate and free it. Don't use the stack for large structs.

jcanseco replied:
Done.

joelagnel commented on src/bpfd.c:
```diff
@@ -331,12 +339,54 @@ int bpf_clear_map(int map_fd, int klen)
+	char *exe_path = get_pid_exe(pid);
+	if (!usym_cache_ || usym_cache_->pid != pid || strcmp(usym_cache_->exe_path, exe_path)) {
+		if (!usym_cache_) {
+			usym_cache_ = (struct usym_cache *)malloc(sizeof(struct usym_cache));
```
why aren't you free it at the end of the program?

jcanseco replied:
Done.

joelagnel commented on src/bpfd.c:
```diff
@@ -331,12 +339,54 @@ int bpf_clear_map(int map_fd, int klen)
+{
+	const int PATHBUF_SIZE = 4096;
+
+	char *exe_path = (char *)malloc(PATHBUF_SIZE);
```
Where are you freeing all the exe_path at the end of the program?

jcanseco replied:
Done.

jcanseco commented:
Made the requested changes. PTAL.
